label_studio/data_import/uploader.py

```diff
@@ -133,7 +133,13 @@ def tasks_from_url(file_upload_ids, project, user, url, could_be_tasks_list):
         validate_upload_url(url, block_local_urls=settings.SSRF_PROTECTION_ENABLED)
         # Reason for #nosec: url has been validated as SSRF safe by the
         # validation check above.
-        response = requests.get(url, verify=False, headers={'Accept-Encoding': None})  # nosec
+        response = requests.get(url, verify=False, stream=True, headers={'Accept-Encoding': None})  # nosec
+
+        # second check for SSRF for prevent redirect and dns rebinding attacks
+        if settings.SSRF_PROTECTION_ENABLED:
+            response_ip = response.raw._connection.sock.getpeername()[0]
+            validate_ip(response_ip)
```

Comment:
This looks a bit too low-level of an implementation to expose in app code; I think it should live in a library function that wraps requests.get and performs validation first on the address, then makes the request, then revalidates the ip on the response, before returning the response object.

I googled around for tools we could use to help us get SSRF safety right and came across this library - it's a real shame that it's archived: https://github.com/JordanMilne/Advocate.

Lastly, did you audit for any other usages of `validate_upload_url` that should be checked for rebinding?

Reply:
A separate ticket created for other usages of `validate_upload_url`: LEAP-233